0000-stream/0000-stream.md

```diff
@@ -0,0 +1,182 @@
+
+| Field | Type | Description |
+|---|---|---|
+| Version | UInt8 | `1` for this version |
+| ILP Packet Type | UInt8 | ILPv4 packet type this STREAM packet MUST be sent in (`12` for Prepare, `13` for Fulfill, and `14` for Reject). A sender or receiver MUST discard a STREAM packet that comes in on the wrong ILP Packet Type. |
+| Sequence | VarUInt | Identifier for an ILP request / resopnse. Clients and Servers track their own outgoing packet sequence numbers and increment the `Sequence` for each ILP Prepare they send. The Receiver MUST respond with a STREAM packet that includes the same `Sequence` as the Sender's Prepare packet. A sender MUST discard a STREAM packet in which the `Sequence` does not match the STREAM packet sent with their ILP Prepare. |
+| Prepare Amount | VarUInt | If the STREAM packet is sent on an ILP Prepare, this represents the minimum the receiver should accept. If the packet is sent on an ILP Fulfill or Reject, this represents the amount that the receiver got in the Prepare. |
+| Frames | SEQUENCE OF Frame | The rest of the packet is comprised of type- and length-prefixed Frames, as specified below. Note that the array of frames is NOT length-prefixed. |
+
+**TODO**: Should we length-prefix the frames array and make the padding a field in the packet format to better comply with OER? Note that we cannot length-prefix the padding because in certain edge cases it could make it impossible to pad to the right length when the length prefix is included.
+
+### Frames and Frame Types
+
+Every frame is encoded with the type and length prefix as follows.
+
+A receiver that gets an unknown frame type SHOULD ignore and skip it.
+
+#### Frame Encoding
+
+See the [ASN.1 definition](../asn1/Stream.asn) for the formal specification.
+
+| Field | Type | Description |
+|---|---|---|
+| Type | UInt8 | Identifier for the frame type (see below) |
+| Data | Variable-Length Octet String | Frame contents |
+
+#### Frame Types
+
+| Type ID | Frame |
+|---|---|
+| `0x01` | Connection New Address |
+| `0x02` | Connection Close |
+| `0x03` | Application Close |
+| `0x04` | Connection Max Money |
+| `0x05` | Connection Money Blocked |
+| `0x06` | Connection Max Data |
+| `0x07` | Connection Data Blocked |
+| `0x08` | Connection Max Stream ID |
+| `0x09` | Connection Stream ID Blocked |
+| `0x10` | Stream Money |
+| `0x11` | Stream Money End |
+| `0x12` | Stream Money Max |
+| `0x13` | Stream Money Blocked |
+| `0x14` | Stream Money Close |
+| `0x20` | Stream Data |
+| `0x21` | Stream Data End |
+| `0x22` | Stream Data Max |
+| `0x23` | Stream Data Blocked |
+| `0x24` | Stream Data Close |
+
+#### Padding
+
+Zero (hex `0x00`) bytes MAY be appended after the other frames, for example to ensure that all packets are the same size to minimize the information gained from passive packet analysis. A parser SHOULD stop reading once it reaches a zero byte instead of a Frame Type Identifier.
```

Comment:
There are really two questions here:
1. Is the padding frame just a single zero byte (0x00, like [Padding in QUIC](https://quicwg.github.io/base-drafts/draft-ietf-quic-transport.html#frame-padding)) or a proper frame that has a length prefix for the contents?
2. Can padding come between other frames (see https://github.com/quicwg/base-drafts/issues/158) or does it have to come at the end?

The problem I ran into with length-prefixing the padding is that OER length prefixes are themselves variable length. You need to know how much padding there is to know how long the length prefix is, but the length of the length prefix affects how much padding you need. I'm not completely sure but I think there is a possible edge case where it's impossible to get the right amount of padding because of that.

If you don't length-prefix the padding, you definitely want to mandate that padding should come at the end, because checking individual bytes is very slow. I ran into this when adding padding to the implementation and at first the padded packets made the tests run thousands of times slower than the tests for unpadded packets. I thought it might be encryption related at first, but it turned out that it was just the speed of checking the value of each byte to see if it was zero (there were >30k of them).